fenix/app/src/main/java/org/mozilla/fenix/components/toolbar/navbar/NavigationBarButtons.kt

```diff
@@ -0,0 +1,206 @@
+import org.mozilla.fenix.tabstray.Page
+
+/**
+ * Navigation bar button types.
+ */
+sealed class NavBarButton {
+    /**
+     * A general action button, like Home or Back buttons.
+     */
+    data class ActionButton(val iconId: Int, val descriptionResourceId: Int) : NavBarButton() {
+        var onClick: () -> Unit = {}
+        var onLongPress: () -> Unit = {}
+    }
+
+    /**
+     * A tab counter.
+     */
+    object TabCounterButton : NavBarButton() {
+        var onClick: () -> Unit = {}
+    }
+
+    /**
+     * Three dot menu button.
+     */
+    data class ThreeDotMenuButton(val menuButton: MenuButton) : NavBarButton()
+}
+
+/**
+ * Holds a collection of default buttons.
+ *
+ * Items could be copied and customized for usage across the app.
+ * Also used for building previews.
+ */
+object DefaultButtons {
+    val home = NavBarButton.ActionButton(
+        iconId = R.drawable.mozac_ic_home_24,
+        descriptionResourceId = R.string.browser_toolbar_home,
+    )
+
+    val search = NavBarButton.ActionButton(
+        iconId = R.drawable.mozac_ic_search_24,
+        descriptionResourceId = R.string.browser_toolbar_home,
+    )
+
+    val menu = NavBarButton.ActionButton(
+        iconId = R.drawable.mozac_ic_ellipsis_vertical_24,
+        descriptionResourceId = R.string.mozac_browser_menu_button,
+    )
+
+    val back = NavBarButton.ActionButton(
+        iconId = R.drawable.mozac_ic_back_24,
+        descriptionResourceId = R.string.browser_menu_back,
+    )
+
+    val forward = NavBarButton.ActionButton(
+        iconId = R.drawable.mozac_ic_forward_24,
+        descriptionResourceId = R.string.browser_menu_forward,
+    )
+
+    val tabs = NavBarButton.TabCounterButton
+
+    val defaultItems = listOf(back, forward, home, tabs, menu)
+}
+
+/**
+ * Defines navigation bar locations with required dependencies for building the bav bar.
+ */
+sealed class NavBarLocation {
+
+    /**
+     * Represents BrowserFragment.
+     */
+    data class BrowserFragment(
+        val interactor: BrowserToolbarInteractor,
+        val navController: NavController,
+        val thumbnailsFeature: ViewBoundFeatureWrapper<BrowserThumbnails>,
+        val browsingModeManager: BrowsingModeManager,
+        val menuButton: MenuButton,
+    ) : NavBarLocation()
+
+    /**
+     * Represents HomeFragment.
+     */
+    data class HomeFragment(
+        val navController: NavController,
+        val browsingModeManager: BrowsingModeManager,
+        val menuButton: MenuButton,
+    ) : NavBarLocation()
+}
+
+/**
+ * Constructs a list of button for a given navigation target.
+ *
+ * This function takes a [NavBarLocation] instance, identifying the current UI context,
+ * and returns a tailored list of [NavBarButton]s suitable for that location. Each [NavBarButton]
+ * is configured based on the specific requirements and characteristics of the navigation target,
+ * ensuring a context-appropriate user interface.
+ */
+object ActionItemListBuilder {
```

Comment:
Reading through this file, it looks like we're trying to simplify some of the API of dynamically creating the navigation bar buttons.

What led you to the creation of `ActionItemListBuilder` + `DefaultButtons` objects? It looks like there's a ton of complexity between those objects, the `NavBarLocation`, and the two transform functions to generate a list of buttons. 

Essentially, I'm having a hard time understanding what problem we're ultimately trying to solve. Based on the code here and the figma, we need to have two versions of the navigation bar (one for browser and one for homescreen), but I have a concern we're coding for significantly beyond the scope of those two use cases.

Reply:
You got it right, I was trying to simplify things.

`DefaultButtons` allow us not create same classes with same resources every time we need to add buttons to a nav bar (we do that twice for regular tabs, twice for private and once for custom). 
`ActionItemListBuilder` just encapsulates the logic of handling the clicks. It's verbose (in case of BrowserFragment it is 50 of lines), and I didn't want to put into the Fragment itself.

It looks like I was trying to solve two problems:
1) not to create buttons with same resources multiple times
2) not putting click handling code into the fragment directly

Reply:
Do you have any concerns with the maintenance of this code? If UX decides to change the layout down the road (whether that's to change the icons/functionality or add a non-icon action to the bar), will the proposed approach still hold up? Can we boost our confidence with the button generation with some tests?

Reply:
yeah, I think it's not a problem. it might be even easier to maintain since it's dynamic. 

but I am not sure we should go against the existing patters we already have, so right let's focus on non-dynamic approach.